### PR TITLE
Add key-value store utility module

### PR DESCRIPTION
**Pull request by jmorrell**
This pulls in the key value store and unit tests that has been prototyped in the Node buildpack into the stdlib: https://github.com/heroku/heroku-buildpack-nodejs/blob/master/lib/kvstore.sh

My attempt at porting the unit tests to bats was unhappy, so I'm proposing either keeping two unit test suites or I can port the existing one to `shunit2` which is what the official buildpacks use anyway.

This is the foundation for several other modules that will come in subsequent PRs

### CODE REVIEW

**Review by jkutner — APPROVED** [review]
I didn't look *extremely* closely at the bash-isms in the `kv_` functions, but given that they don't affect existing behavior, and only get pulled in on-demand, I'm fine merging.

**Comment by jmorrell** [comment]
@dzuelke would you like to review the bash-isms here? This is already pretty tested out in production, but I welcome any feedback

File: stdlib.sh [hunk 1]
```diff
@@ -61,6 +61,113 @@ puts_verbose() {
...
+  local f=$1
+  mkdir -p "$(dirname "$f")"
+  touch "$f"
+}
+
+# Delete all keys and values from the store
+# Usage: kv_clear file
+kv_clear() {
+  local f=$1
+  echo "" > "$f"
```
Comment (dzuelke) [inline stdlib.sh:102]: this can be just `: > "$f"`

File: stdlib.sh [hunk 2]
```diff
@@ -61,6 +61,113 @@ puts_verbose() {
...
+}
+
+# Set value for a key
+# Usage: kv_set file key value
+kv_set() {
+  if [[ $# -eq 3 ]]; then
```
Comment (dzuelke) [inline stdlib.sh]: should this raise an error if args incorrect?

Comment (dzuelke) [reply stdlib.sh]: also, you can make it `if (( $# == 3 )); then`, or, if you want to handle errors, `if (( $# != 3 )); then`, and handle and throw the error in there, so the rest of the function isn't indented

Comment (dzuelke) [reply stdlib.sh]: in general, as a style, I prefer an `if` with the negated condition that just does a `return` (or throws error or whatever) over wrapping the entire function block in the `if`

File: stdlib.sh [hunk 3]
```diff
@@ -61,6 +61,113 @@ puts_verbose() {
...
+    local f=$1
+    if [[ -f $f ]]; then
```
Comment (dzuelke) [inline stdlib.sh]: `-w` is better here, checks for writeability

File: stdlib.sh [hunk 4]
```diff
@@ -61,6 +61,113 @@ puts_verbose() {
...
+      echo "$2=$3" >> "$f"
+    fi
+  fi
+}
+
+# Retrieve value for a key. Will return an empty string if the key is
+# not present
+# Usage: kv_get file key
+kv_get() {
+  if [[ $# -eq 2 ]]; then
```
Comment (dzuelke) [inline stdlib.sh]: again, `if (( … ))` and maybe the indent thing

File: stdlib.sh [hunk 5]
```diff
@@ -61,6 +61,113 @@ puts_verbose() {
...
+    local f=$1
+    if [[ -f $f ]]; then
```
Comment (dzuelke) [inline stdlib.sh]: you could use `-r` here

File: stdlib.sh [hunk 6]
```diff
@@ -61,6 +61,113 @@ puts_verbose() {
...
+}
+
+# Log out each of the keys and values in the store, one per line
+# Usage: kv_list file
+kv_list() {
+  local f=$1
+
+  kv_keys "$f" | tr ' ' '\n' | while read -r key; do
+    if [[ -n $key ]]; then
+      echo "$key=$(kv_get_escaped "$f" "$key")"
```
Comment (dzuelke) [inline stdlib.sh:191]: you could maybe also use `printf %q "$value"` instead of this escaping stuff

File: stdlib.sh [hunk 7]
```diff
@@ -61,6 +61,113 @@ puts_verbose() {
...
+  echo "" > "$f"
+}
+
+# Set value for a key
+# Usage: kv_set file key value
+kv_set() {
+  if [[ $# -eq 3 ]]; then
+    local f=$1
+    if [[ -f $f ]]; then
+      echo "$2=$3" >> "$f"
```
Comment (dzuelke) [inline stdlib.sh]: this breaks with newlines in `$3`, so I think `printf "$2=%q\n" "$3"` is better here; that stores it as escaped

File: stdlib.sh [hunk 8]
```diff
@@ -61,6 +61,113 @@ puts_verbose() {
...
+    fi
+  fi
+}
+
+# Retrieve value for a key. Will return an empty string if the key is
+# not present
+# Usage: kv_get file key
+kv_get() {
```
Comment (dzuelke) [inline stdlib.sh:135]: then maybe change this function to `kv_get_raw`, which just returns the line like it already does...

File: stdlib.sh [hunk 9]
```diff
@@ -61,6 +61,113 @@ puts_verbose() {
...
+    if [[ -f $f ]]; then
+      grep "^$2=" "$f" | sed -e "s/^$2=//" | tail -n 1
+    fi
+  fi
+}
+
+# Retrive the value for a key, but wrap it in quotes if the value contains spaces
+# This is expected to be mostly used internally
+# Usage: kv_get_escaped file key
+kv_get_escaped() {
```
Comment (dzuelke) [inline stdlib.sh:149]: ... and make this `kv_get`, which does `printf "%b\n" "$(kv_get_raw …)"` (the quoting might be off there)

**Review by dzuelke — COMMENTED** [review]
Leaving some comments mostly around spaces in values; I also think we should validate keys to match `[a-zA-Z_][a-zA-Z0-9_]+` for forward compat (who knows what we want to do with this stuff at some point, think variables extraction or whatever, and this is a sane rule for key/variable names in basically every language)

File: stdlib.sh (hunk 1 above)
Comment (jmorrell) [reply stdlib.sh:102]: I find that less clear so I'm going to keep it as-is here

File: stdlib.sh (hunk 6 above)
Comment (jmorrell) [reply stdlib.sh:191]: I tried this, but could not get it to unescape with `printf %b`

File: stdlib.sh (hunk 7 above)
Comment (jmorrell) [reply stdlib.sh]: I went down this path, but couldn't make everything work the way I wanted. In the end I decided to flatten any multiline strings.

Most of my hesitation here comes from the likely need to parse these files in go or ruby soon and wanting to keep the escaping rules very simple

File: test/unit [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+  assertEquals "" "$(kv_get "$store" key)"
+  assertEquals "" "$(kv_keys "$store")"
+  assertEquals "" "$(kv_list "$store")"
+
+  # running these commands has not created this file
+  assertTrue "[[ ! -e $store ]]"
+}
+
+testKeyValueMultiLineString() {
```
Comment (jmorrell) [inline test/unit:85]: New unit tests start here

**Review by dzuelke — APPROVED** [review]

**Pull request merged**